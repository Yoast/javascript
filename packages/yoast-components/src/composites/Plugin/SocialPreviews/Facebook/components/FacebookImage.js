/* External dependencies */
import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { __ } from "@wordpress/i18n";

/* Yoast dependencies. */
import { colors } from "@yoast/style-guide";

/* Internal dependencies */
import {
	determineFacebookImageProperties,
	LANDSCAPE_HEIGHT,
	LANDSCAPE_WIDTH,
	PORTRAIT_HEIGHT,
	PORTRAIT_WIDTH,
	SQUARE_HEIGHT,
	SQUARE_WIDTH,
} from "../helpers/determineFacebookImageProperties";

const MIN_IMAGE_WIDTH = 158;
const MIN_IMAGE_HEIGHT = 158;

const FacebookImageContainer = styled.div`
	position: relative;
	height: ${ props => props.dimensions.height };
	width: ${ props => props.dimensions.width };
	overflow: hidden;
	background-color: ${ colors.$color_white };
`;

const StyledImage = styled.img`
	width: ${ props => props.imageProperties.width }px;
	height: ${ props => props.imageProperties.height }px;
	position: absolute;
	top: 50%;
	left: 50%;
	transform: translate(-50%, -50%);
`;

const ErrorImage = styled.p`
	display: flex;
	justify-content: center;
	align-items: center;
	box-sizing: border-box;
	width: 500px;
	height: 261px;
	max-width: 100%;
	margin: 0;
	padding: 1em;
	text-align: center;
	font-size: 1rem;
	color: ${ colors.$color_white };
	background-color: ${ colors.$color_red };
`;

const PlaceholderImage = styled.div`
	height: 261px;
	width: 500px;
	background-color: ${ colors.$color_grey };
`;

/**
 * Renders the FacebookImage component.
 *
 * @param {string} src The image source.
 *
 * @returns {ReactComponent} The FacebookImage component.
 */
export default class FacebookImage extends React.Component {
	/**
	 * The constructor.
	 *
	 * @param {Object} props The component's props.
	 */
	constructor( props ) {
		super( props );
		this.state = {
			imageProperties: null,
			status: "loading",
		};
	}

	/**
	 * After the component did mount, determine the properties of the FacebookImage.
	 *
	 * @returns {Promise} Resolves when there are image properties.
	 */
	componentDidMount() {
		return determineFacebookImageProperties( this.props.src ).then( ( imageProperties ) => {
			this.setState( {
				imageProperties: imageProperties,
				status: "loaded",
			} );
		} ).catch( () => {
			this.setState( {
				imageProperties: null,
				status: "errored",
			} );
			return true;
		} );
	}

	/**
	 * Gets the dimensions for the Facebook image container.
	 *
	 * @param {string} imageMode The facebook image mode: either landscape, square or portrait.
	 *
	 * @returns {Object} The width and height for the container.
	 */
	getContainerDimensions( imageMode ) {
		switch ( imageMode ) {
			case "square":
				return {
					height: SQUARE_HEIGHT + "px",
					width: SQUARE_WIDTH + "px",
				};
			case "portrait":
				return {
					height: PORTRAIT_HEIGHT + "px",
					width: PORTRAIT_WIDTH + "px",
				};
			case "landscape":
			default:
				return {
					height: LANDSCAPE_HEIGHT + "px",
					width: LANDSCAPE_WIDTH + "px",
				};
		}
	}

	/**
	 * Renders the FacebookImage.
	 *
	 * @returns {ReactComponent} Either the ErrorImage component or the FacebookImageContainer.
	 */
	render() {
		const imageProperties = this.state.imageProperties;
		const status = this.state.status;

		if ( status === "loading" ) {
			return <PlaceholderImage />;
		}

		if ( status === "errored" ) {
			return <ErrorImage>{ __( "The given image url cannot be loaded", "yoast-components" ) }</ErrorImage>;
		}

		if ( imageProperties.height < MIN_IMAGE_HEIGHT || imageProperties.width < MIN_IMAGE_WIDTH ) {
			return <ErrorImage>{ __( "The image you selected is too small for Facebook", "yoast-components" ) }</ErrorImage>;
		}

		const containerDimensions = this.getContainerDimensions( imageProperties.mode );
		return <FacebookImageContainer
			dimensions={ containerDimensions }
		>
			<StyledImage
				src={ this.props.src }
				alt={ this.props.alt }
				imageProperties={ imageProperties }
			/>
		</FacebookImageContainer>;
	}
}

FacebookImage.propTypes = {
	src: PropTypes.string.isRequired,
	alt: PropTypes.string,
};

FacebookImage.defaultProps = {
	alt: "",
};
