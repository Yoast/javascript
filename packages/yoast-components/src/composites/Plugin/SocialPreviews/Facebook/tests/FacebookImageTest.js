/* External dependencies */
import React from "react";
import renderer from "react-test-renderer";

/* Internal dependencies */
import FacebookImage from "../components/FacebookImage";
import delayComponentSnapshot from "./testHelpers/delayComponentSnapshot";

const importedDetermineFacebookImageProperties = require( "../helpers/determineFacebookImageProperties.js" );

jest.mock( "../helpers/determineFacebookImageProperties.js", () => {
	return {
		determineFacebookImageProperties: jest.fn(),
		SQUARE_WIDTH: 158,
		SQUARE_HEIGHT: 158,
		PORTRAIT_WIDTH: 158,
		PORTRAIT_HEIGHT: 236,
		LANDSCAPE_WIDTH: 500,
		LANDSCAPE_HEIGHT: 261,
	};
} );

describe( "FacebookImage Component", () => {
	it( "calls determineFacebookImageProperties with the correct image URL", () => {
		const imageUrl = "https://yoast.com/app/uploads/2015/09/Author_Joost_x2.png";

		importedDetermineFacebookImageProperties.determineFacebookImageProperties.mockReturnValueOnce( Promise.resolve( {
			mode: "portrait",
			height: 600,
			width: 300,
		} ) );

		renderer.create(
			<FacebookImage src={ imageUrl } />
		);

		expect( importedDetermineFacebookImageProperties.determineFacebookImageProperties ).toBeCalledWith( imageUrl );
	} );

	it( "matches the snapshot in the loading state", () => {
		importedDetermineFacebookImageProperties.determineFacebookImageProperties.mockReturnValueOnce(
			// Make sure the promise is never resolved to keep the loading state.
			new Promise( () => {} )
		);
		const component = renderer.create(
			<FacebookImage src="https://yoast.com/app/uploads/2015/09/Author_Joost_x2.png" />
		);
		const tree = component.toJSON();
		expect( tree ).toMatchSnapshot();
	} );

	it( "matches the snapshot for a portrait image", () => {
		importedDetermineFacebookImageProperties.determineFacebookImageProperties.mockReturnValueOnce( Promise.resolve( {
			mode: "portrait",
			height: 600,
			width: 300,
		} ) );
		const component = renderer.create(
			<FacebookImage src="https://yoast.com/app/uploads/2015/09/Author_Joost_x2.png" />
		);

		// Wait for the determineFacebookImageProperties promise to resolve.
		return delayComponentSnapshot( component );
	} );

	it( "matches the snapshot for a landscape image", () => {
		importedDetermineFacebookImageProperties.determineFacebookImageProperties.mockReturnValueOnce( Promise.resolve( {
			mode: "landscape",
			height: 300,
			width: 600,
		} ) );
		const component = renderer.create(
			<FacebookImage src="https://yoast.com/app/uploads/2015/06/How_to_choose_keywords_FI.png" />
		);

		// Wait for the determineFacebookImageProperties promise to resolve.
		return delayComponentSnapshot( component );
	} );

	it( "matches the snapshot for a square image", () => {
		importedDetermineFacebookImageProperties.determineFacebookImageProperties.mockReturnValueOnce( Promise.resolve( {
			mode: "square",
			height: 300,
			width: 300,
		} ) );
		const component = renderer.create(
			<FacebookImage src="https://yoast.com/app/uploads/2018/09/avatar_user_1_1537774226.png" />
		);

		// Wait for the determineFacebookImageProperties promise to resolve.
		return delayComponentSnapshot( component );
	} );

	it( "matches the snapshot for a too small image", () => {
		importedDetermineFacebookImageProperties.determineFacebookImageProperties.mockReturnValue( Promise.resolve( {
			mode: "square",
			height: 100,
			width: 100,
		} ) );
		const component = renderer.create(
			<FacebookImage src="https://yoast.com/app/uploads/2018/11/Logo_TYPO3-250x105.png" />
		);

		// Wait for the determineFacebookImageProperties promise to resolve.
		return delayComponentSnapshot( component );
	} );

	it( "matches the snapshot for a faulty image", () => {
		importedDetermineFacebookImageProperties.determineFacebookImageProperties.mockReturnValueOnce( Promise.reject() );

		const component = renderer.create(
			<FacebookImage src="thisisnoimage" />
		);

		// Wait for the determineFacebookImageProperties promise to resolve.
		return delayComponentSnapshot( component );
	} );
} );

describe( "getContainerDimensions", () => {
	it( "gets the container dimensions for a landscape image.", () => {
		const FacebookImageComponent = new FacebookImage();

		const actual = FacebookImageComponent.getContainerDimensions( "landscape" );
		const expected = { height: "261px", width: "500px" };

		expect( actual ).toEqual( expected );
	} );

	it( "gets the container dimensions for a square image.", () => {
		const FacebookImageComponent = new FacebookImage();

		const actual = FacebookImageComponent.getContainerDimensions( "square" );
		const expected = { height: "158px", width: "158px" };

		expect( actual ).toEqual( expected );
	} );

	it( "gets the container dimensions for a portrait image.", () => {
		const FacebookImageComponent = new FacebookImage();

		const actual = FacebookImageComponent.getContainerDimensions( "portrait" );
		const expected = { height: "236px", width: "158px" };

		expect( actual ).toEqual( expected );
	} );
} );
