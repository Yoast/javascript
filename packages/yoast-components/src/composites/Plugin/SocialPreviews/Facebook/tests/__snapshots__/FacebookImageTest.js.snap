// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FacebookImage Component matches the snapshot for a faulty image 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  width: 500px;
  height: 261px;
  max-width: 100%;
  margin: 0;
  padding: 1em;
  text-align: center;
  font-size: 1rem;
  color: #fff;
  background-color: #dc3232;
}

<p
  className="c0"
>
  The given image url cannot be loaded
</p>
`;

exports[`FacebookImage Component matches the snapshot for a landscape image 1`] = `
.c0 {
  position: relative;
  height: 261px;
  width: 500px;
  overflow: hidden;
  background-color: #fff;
}

.c1 {
  width: 600px;
  height: 300px;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

<div
  className="c0"
>
  <img
    alt=""
    className="c1"
    src="https://yoast.com/app/uploads/2015/06/How_to_choose_keywords_FI.png"
  />
</div>
`;

exports[`FacebookImage Component matches the snapshot for a portrait image 1`] = `
.c0 {
  position: relative;
  height: 236px;
  width: 158px;
  overflow: hidden;
  background-color: #fff;
}

.c1 {
  width: 300px;
  height: 600px;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

<div
  className="c0"
>
  <img
    alt=""
    className="c1"
    src="https://yoast.com/app/uploads/2015/09/Author_Joost_x2.png"
  />
</div>
`;

exports[`FacebookImage Component matches the snapshot for a square image 1`] = `
.c0 {
  position: relative;
  height: 158px;
  width: 158px;
  overflow: hidden;
  background-color: #fff;
}

.c1 {
  width: 300px;
  height: 300px;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

<div
  className="c0"
>
  <img
    alt=""
    className="c1"
    src="https://yoast.com/app/uploads/2018/09/avatar_user_1_1537774226.png"
  />
</div>
`;

exports[`FacebookImage Component matches the snapshot for a too small image 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  width: 500px;
  height: 261px;
  max-width: 100%;
  margin: 0;
  padding: 1em;
  text-align: center;
  font-size: 1rem;
  color: #fff;
  background-color: #dc3232;
}

<p
  className="c0"
>
  The image you selected is too small for Facebook
</p>
`;

exports[`FacebookImage Component matches the snapshot in the loading state 1`] = `
.c0 {
  height: 261px;
  width: 500px;
  background-color: #ddd;
}

<div
  className="c0"
/>
`;
